File: src/kaniko_remote/k8s/specs.py
from __future__ import annotations

from typing import List

from kubernetes.client.models import (
    V1ConfigMapEnvSource,
    V1Container,
    V1EnvFromSource,
    V1EnvVar,
    V1Pod,
    V1PodSpec,
    V1SecretReference,
    V1SecretVolumeSource,
    V1Volume,
    V1VolumeMount,
)

from kaniko_remote import __version__


class K8sSpecs:
    @classmethod
    def generate_pod_spec(
        cls,
        name: str,
        cpu: str,
        memory: str,
        kaniko_image: str,
        setup_image: str,
        additional_labels: dict,
        additional_annotations: dict,
        **kwargs,
    ) -> V1Pod:
        docker_config_volume_mounts = [V1VolumeMount(name="config", mount_path="/kaniko/.docker")]

        resources = dict(
            limits=dict(cpu=cpu, memory=memory),
            requests=dict(cpu=cpu, memory=memory),
        )

        # Additional labels/annotations could actually be a DeflatableDict at this point,
        # which openapi-generator throws up about, so force each into dict before adding
        # 'system' labels/annotations
        labels = {
            **additional_labels,
            **{
                "app.kubernetes.io/name": "kaniko-remote",
                "app.kubernetes.io/component": "builder",
                "kaniko-remote/builder-name": name,
            },
        }
        annotations = {
            **additional_annotations,
            **{
                "kaniko-remote/version": __version__,
            },
        }

        return V1Pod(
            api_version="v1",
            kind="Pod",
            metadata=dict(
                generateName=f"kaniko-remote-{name}-",
                labels=labels,
                annotations=annotations,
            ),
            spec=V1PodSpec(
                automount_service_account_token=False,
                init_containers=[
                    V1Container(
                        name="setup",
                        image=setup_image,
                        command=["sh", "-c"],
                        # args=["trap : TERM INT; sleep 9999999999d & wait"],
                        args=["until [ -e /kaniko/.docker/config.json ]; do sleep 1; done"],
                        volume_mounts=docker_config_volume_mounts,
                        resources=resources,
                    )
                ],
                containers=[
                    V1Container(
                        name="builder",
                        image=kaniko_image,
                        # args=kaniko_args,
                        # command=["/busybox/sh", "-c"],
                        # args=["trap : TERM INT; sleep 9999999999d & wait"],
                        volume_mounts=docker_config_volume_mounts,
                        resources=resources,
                    )
                ],
                volumes=[V1Volume(name="config", empty_dir={})],
            ),
        )

    @classmethod
    def mount_context_for_exec_transfer(cls, pod: V1Pod) -> V1Pod:
        pod.spec.volumes.append(V1Volume(name="context", empty_dir={}))

        # For some reason appending to either the main container or the init container appends to both,
        # so we only append to one.
        pod.spec.containers[0].volume_mounts.append(V1VolumeMount(name="context", mount_path="/workspace"))
        return pod

    @classmethod
    def set_kaniko_args(cls, pod: V1Pod, preparsed_args: List[str], **kwargs) -> V1Pod:
        required_args = ["destinations"]

        for required_arg in required_args:
            if required_arg not in kwargs:
                raise ValueError(f"Missing required kaniko argument --{required_arg}")

        # Start with defaults which may be overriden
        args = {
            "dockerfile": "Dockerfile",
            "digest-file": "/dev/termination-log",
        }

        # Pop multi-args so they can be expanded later
        destinations = kwargs.pop("destinations")
        build_args = kwargs.pop("build_args", ())
        labels = kwargs.pop("labels", ())

        args.update(kwargs)
        args = [f"--{k}={v}" for k, v in args.items() if v is not None]

        args.extend(preparsed_args)

        for d in destinations:
            args.append(f"--destination={d}")
        for ba in build_args:
            args.append("--build-arg")
            args.append(ba)
        for l in labels:
            args.append(f"--label={l}")

        pod.spec.containers[0].command = None
        pod.spec.containers[0].args = args
        return pod

    @classmethod
    def replace_service_account(cls, pod: V1Pod, service_account_name: str) -> V1Pod:
        pod.spec.service_account_name = service_account_name
        pod.spec.automount_service_account_token = True
        return pod

    @classmethod
    def append_env_from_secret(cls, pod: V1Pod, secret_name: str) -> V1Pod:
        container: V1Container = pod.spec.containers[0]
        if not container.env_from:
            container.env_from = []
        container.env_from.append(V1EnvFromSource(secret_ref=V1SecretReference(name=secret_name)))
        return pod

    @classmethod
    def append_env_from_config_map(cls, pod: V1Pod, config_map_name: str) -> V1Pod:
        container: V1Container = pod.spec.containers[0]
        if not container.env_from:
            container.env_from = []
        container.env_from.append(V1EnvFromSource(config_map_ref=V1ConfigMapEnvSource(name=config_map_name)))
        return pod

    @classmethod
    def append_env_var(cls, pod: V1Pod, env_var_name: str, env_var_value) -> V1Pod:
        container: V1Container = pod.spec.containers[0]
        if not container.env:
            container.env = []
        container.env.append(V1EnvVar(name=env_var_name, value=env_var_value))
        return pod

    @classmethod
    def append_volume_from_secret(cls, pod: V1Pod, secret_name: str, mount_path: str) -> V1Pod:
        mounts: List[V1VolumeMount] = pod.spec.containers[0].volume_mounts
        volumes: List[V1Volume] = pod.spec.volumes
        mounts.append(V1VolumeMount(name=secret_name, mount_path=mount_path))
        volumes.append(V1Volume(name=secret_name, secret=V1SecretVolumeSource(secret_name=secret_name)))
        return pod

    @classmethod
    def append_volume_from_config_map(cls, pod: V1Pod, config_map_name: str, mount_path: str) -> V1Pod:
        mounts: List[V1VolumeMount] = pod.spec.containers[0].volume_mounts
        volumes: List[V1Volume] = pod.spec.volumes
        mounts.append(V1VolumeMount(name=config_map_name, mount_path=mount_path))
        volumes.append(V1Volume(name=config_map_name, config_map=V1ConfigMapEnvSource(name=config_map_name)))
        return pod

File: src/K8s/Specs.cs
using k8s.Models;

namespace KanikoRemote.K8s
{
    internal static class Specs
    {
        public static V1Pod GeneratePodSpec(
            string name,
            string cpu,
            string memory,
            string kanikoImage,
            string setupImage,
            IDictionary<string, string> additionalLabels,
            IDictionary<string, string> additionalAnnotations)
        {
            var dockerConfigVolumeMounts = new List<V1VolumeMount>()
            {
                new V1VolumeMount
                {
                    Name = "config",
                    MountPath = "/kaniko/.docker"
                }
            };

            var resourcesDict = new Dictionary<string, ResourceQuantity>()
            {
                { "cpu", new ResourceQuantity(cpu) },
                { "memory", new ResourceQuantity(memory) }
            };
            var resources = new V1ResourceRequirements()
            {
                Requests = resourcesDict,
                Limits = resourcesDict,
            };

            var labels = new Dictionary<string, string>(additionalLabels);
            labels["app.kubernetes.io/name"] = "kaniko-remote";
            labels["app.kubernetes.io/component"] = "builder";
            labels["kaniko-remote/builder-name"] = name;

            var annotations = new Dictionary<string, string>(additionalAnnotations);
            // TODO:
            // labels["kanaiko-remote/version"] = __version__;

            return new V1Pod
            {
                Metadata = new V1ObjectMeta
                {
                    GenerateName = $"kaniko-remote-{name}-",
                    Labels = labels,
                    Annotations = annotations,
                },
                Spec = new V1PodSpec
                {
                    AutomountServiceAccountToken = false,
                    InitContainers = new List<V1Container>()
                    {
                        new V1Container
                        {
                            Name = "setup",
                            Image = setupImage,
                            Command = new List<string> { "sh", "-c" },
                            // Args = new List<string> { "trap : TERM INT; sleep 9999999999d & wait" },
                            Args = new List<string> { "until [ -e /kaniko/.docker/config.json ]; do sleep 1; done" },
                            VolumeMounts = dockerConfigVolumeMounts,
                            Resources = resources,
                        }
                    },
                    Containers = new List<V1Container>()
                    {
                        new V1Container
                        {
                            Name = "builder",
                            Image = kanikoImage,
                            VolumeMounts = dockerConfigVolumeMounts,
                            Resources = resources
                        }
                    },
                    Volumes = new List<V1Volume>()
                    {
                        new V1Volume()
                        {
                            Name = "config",
                            EmptyDir = new V1EmptyDirVolumeSource()
                        }
                    }
                }
            };
        }

        public static V1Pod MountContextForExecTransfer(V1Pod pod)
        {
            pod.Spec.Volumes.Add(new V1Volume()
            {
                Name = "context",
                EmptyDir = new V1EmptyDirVolumeSource()
            });

            pod.Spec.Containers.First().VolumeMounts.Add(new V1VolumeMount()
            {
                Name = "context",
                MountPath = "/workspace"
            });
            return pod;
        }

        public static V1Pod SetKanikoArgs(V1Pod pod, IEnumerable<string> args)
        {
            pod.Spec.Containers.First().Command = null;
            pod.Spec.Containers.First().Args = args.ToList();

            return pod;
        }

        public static V1Pod ReplaceServiceAccount(V1Pod pod, string serviceAccountName)
        {
            pod.Spec.ServiceAccountName = serviceAccountName;
            pod.Spec.AutomountServiceAccountToken = true;
            return pod;
        }

        public static V1Pod AppendEnvFromSecret(V1Pod pod, string secretName)
        {
            var container = pod.Spec.Containers.First();
            if (container.EnvFrom == null)
            {
                container.EnvFrom = new List<V1EnvFromSource>();
            }
            container.EnvFrom.Add(new V1EnvFromSource()
            {
                SecretRef = new V1SecretEnvSource()
                {
                    Name = secretName,
                },
            });
            return pod;
        }

        public static V1Pod AppendEnvFromConfigMap(V1Pod pod, string configMapName)
        {
            var container = pod.Spec.Containers.First();
            if (container.EnvFrom == null)
            {
                container.EnvFrom = new List<V1EnvFromSource>();
            }
            container.EnvFrom.Add(new V1EnvFromSource()
            {
                ConfigMapRef = new V1ConfigMapEnvSource()
                {
                    Name = configMapName,
                }
            });
            return pod;
        }

        public static V1Pod AppendEnvVar(V1Pod pod, string envVarName, string envVarValue)
        {
            var container = pod.Spec.Containers.First();
            if (container.Env == null)
            {
                container.Env = new List<V1EnvVar>();
            }
            container.Env.Add(new V1EnvVar()
            {
                Name = envVarName,
                Value = envVarValue
            });
            return pod;
        }

        public static V1Pod AppendVolumeFromSecret(V1Pod pod, string secretName, string mountPath)
        {
            var mounts = pod.Spec.Containers.First().VolumeMounts;
            var volumes = pod.Spec.Volumes;
            mounts.Add(new V1VolumeMount()
            {
                Name = secretName,
                MountPath = mountPath,
                ReadOnlyProperty = true,
            });
            volumes.Add(new V1Volume()
            {
                Name = secretName,
                Secret = new V1SecretVolumeSource()
                {
                    SecretName = secretName,
                }
            });
            return pod;
        }

        public static V1Pod AppendVolumeFromConfigMap(V1Pod pod, string configMapName, string mountPath)
        {
            var mounts = pod.Spec.Containers.First().VolumeMounts;
            var volumes = pod.Spec.Volumes;
            mounts.Add(new V1VolumeMount()
            {
                Name = configMapName,
                MountPath = mountPath,
                ReadOnlyProperty = true,
            });
            volumes.Add(new V1Volume()
            {
                Name = configMapName,
                ConfigMap = new V1ConfigMapVolumeSource()
                {
                    Name = configMapName
                }
            });
            return pod;
        }
    }
}